### PR TITLE
Add weibull_cdf Presto function

### LINKED ISSUE #4760: Add missing function to handle weibull_cdf function
### Description

add missing function to handle weibull_cdf prestosql function.

### PR DESCRIPTION
**Pull request by Krishna-Prasad-P-V**
Resolves #4760

### CODE REVIEW

**Comment by Krishna-Prasad-P-V** [comment]
@aditi-pandit @majetideepak - Kindly review this pull request. Thanks.

**Comment by Krishna-Prasad-P-V** [comment]
Fuzzer test output for weibull_cdf:
```
I20230628 09:47:29.582652 1731535 ExpressionVerifier.cpp:56] ===================
I20230628 09:47:29.582690 1731535 ExpressionVerifier.cpp:67] All results match.
I20230628 09:47:29.582752 1731535 ExpressionFuzzer.cpp:1357] ==============================> Done with iteration 850
I20230628 09:47:29.583197 1731535 ExpressionFuzzer.cpp:1092] ==============================> Top 1 by number of rows processed
I20230628 09:47:29.583209 1731535 ExpressionFuzzer.cpp:1094] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230628 09:47:29.583216 1731535 ExpressionFuzzer.cpp:1098] weibull_cdf 96259 100.00% 10164
I20230628 09:47:29.583249 1731535 ExpressionFuzzer.cpp:1104] ==============================> Bottom 1 by number of rows processed
I20230628 09:47:29.583256 1731535 ExpressionFuzzer.cpp:1106] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230628 09:47:29.583261 1731535 ExpressionFuzzer.cpp:1111] weibull_cdf 96259 100.00% 10164
I20230628 09:47:29.583276 1731535 ExpressionFuzzer.cpp:1124] ==============================> All stats sorted by number of times the function was chosen
I20230628 09:47:29.583282 1731535 ExpressionFuzzer.cpp:1126] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230628 09:47:29.583292 1731535 ExpressionFuzzer.cpp:1130] weibull_cdf 96259 100.00% 10164
```

File: velox/functions/prestosql/tests/ProbabilityTest.cpp [hunk 1]
```diff
@@ -155,5 +155,20 @@ TEST_F(ProbabilityTest, binomialCDF) {
...
 }
 
+TEST_F(ProbabilityTest, weibullCDF) {
+  const auto weibullCDF = [&](std::optional<double> a,
+                              std::optional<double> b,
+                              std::optional<double> value) {
+    return evaluateOnce<double>("weibull_cdf(c0, c1, c2)", a, b, value);
+  };
+
+  EXPECT_EQ(weibullCDF(1.0, 1.0, 0.0), 0.0);
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/ProbabilityTest.cpp:165]: Please add tests with parameters std::nullopt, kInf, kDoubleMax, kDoubleMin and kNan boundary cases.

File: velox/docs/functions/presto/math.rst [hunk 2]
```diff
@@ -238,6 +238,10 @@ Floating Point Functions
 
     Returns the constant representing not-a-number.
 
+.. function:: weibull_cdf(a, b, value) -> double
```
Comment (aditi-pandit) [inline velox/docs/functions/presto/math.rst:241]: This function should be in the next section with Probability functions.

Comment (Krishna-Prasad-P-V) [reply velox/docs/functions/presto/math.rst:241]: Fixed.

**Comment by Krishna-Prasad-P-V** [comment]
@aditi-pandit - Kindly re-review this pull request. Handled the provided comments. Thanks.

Adding fuzzer test output for reference
```
I20230628 15:03:41.710202 2072104 ExpressionVerifier.cpp:56] ===================
I20230628 15:03:41.710215 2072104 ExpressionVerifier.cpp:67] All results match.
I20230628 15:03:41.710250 2072104 ExpressionFuzzer.cpp:1357] ==============================> Done with iteration 408
I20230628 15:03:41.710522 2072104 ExpressionFuzzer.cpp:1092] ==============================> Top 1 by number of rows processed
I20230628 15:03:41.710542 2072104 ExpressionFuzzer.cpp:1094] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230628 15:03:41.710554 2072104 ExpressionFuzzer.cpp:1098] weibull_cdf 42796 100.00% 5241
I20230628 15:03:41.710570 2072104 ExpressionFuzzer.cpp:1104] ==============================> Bottom 1 by number of rows processed
I20230628 15:03:41.710600 2072104 ExpressionFuzzer.cpp:1106] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230628 15:03:41.710608 2072104 ExpressionFuzzer.cpp:1111] weibull_cdf 42796 100.00% 5241
I20230628 15:03:41.710615 2072104 ExpressionFuzzer.cpp:1124] ==============================> All stats sorted by number of times the function was chosen
I20230628 15:03:41.710621 2072104 ExpressionFuzzer.cpp:1126] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230628 15:03:41.710628 2072104 ExpressionFuzzer.cpp:1130] weibull_cdf 42796 100.00% 5241
```

**Comment by aditi-pandit** [comment]
@Krishna-Prasad-P-V : Please can you check the linux-build failure.

File: velox/functions/prestosql/Probability.h [hunk 3]
```diff
@@ -91,5 +92,25 @@ struct BinomialCDFFunction {
   }
 };
 
+template <typename T>
+struct WeibullCDFFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void
+  call(double& result, double a, double b, double value) {
+    constexpr double kInf = std::numeric_limits<double>::infinity();
```
Comment (aditi-pandit) [inline velox/functions/prestosql/Probability.h:101]: Make this a static variable. static constexpr double kInf = ....

File: velox/functions/prestosql/tests/ProbabilityTest.cpp [hunk 4]
```diff
@@ -155,5 +155,35 @@ TEST_F(ProbabilityTest, binomialCDF) {
...
+  EXPECT_EQ(weibullCDF(1.0, 1.0, 40.0), 1.0);
+  EXPECT_EQ(weibullCDF(1.0, 0.6, 3.0), 0.99326205300091452);
+  EXPECT_EQ(weibullCDF(1.0, 0.9, 2.0), 0.89163197677810413);
+  EXPECT_EQ(weibullCDF(std::nullopt, 1.0, 0.3), std::nullopt);
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/ProbabilityTest.cpp:169]: Nit : Please add a newline before and after the std::nullopt cases.

**Review by aditi-pandit — COMMENTED** [review]
Thanks. This is mostly looking good.

File: velox/functions/prestosql/Probability.h (hunk 3 above)
Comment (Krishna-Prasad-P-V) [reply velox/functions/prestosql/Probability.h:101]: Fixed.

File: velox/functions/prestosql/tests/ProbabilityTest.cpp (hunk 4 above)
Comment (Krishna-Prasad-P-V) [reply velox/functions/prestosql/tests/ProbabilityTest.cpp:170]: Fixed.

File: velox/functions/prestosql/tests/ProbabilityTest.cpp [hunk 5]
```diff
@@ -155,5 +155,39 @@ TEST_F(ProbabilityTest, binomialCDF) {
...
+  EXPECT_EQ(weibullCDF(1.0, 0.6, 3.0), 0.99326205300091452);
+  EXPECT_EQ(weibullCDF(1.0, 0.9, 2.0), 0.89163197677810413);
+
+  EXPECT_EQ(weibullCDF(std::nullopt, 1.0, 0.3), std::nullopt);
+
+  EXPECT_EQ(weibullCDF(1.0, std::nullopt, 0.2), std::nullopt);
+
+  EXPECT_EQ(weibullCDF(1.0, 0.4, std::nullopt), std::nullopt);
+
+  EXPECT_EQ(weibullCDF(kDoubleMin, 1.0, 2.0), 0.63212055882855767);
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/ProbabilityTest.cpp:176]: Just curious : Why are we not testing varied combinations for value parameter ( < 0, kDoubleMin, kDoubleMax, kInf, kNan) etc ?

**Comment by yzhang1991** [comment]
Hi @Krishna-Prasad-P-V, is this ready for another review?

File: velox/functions/prestosql/tests/ProbabilityTest.cpp (hunk 5 above)
Comment (Krishna-Prasad-P-V) [reply velox/functions/prestosql/tests/ProbabilityTest.cpp:240]: Added and tested.

**Comment by Krishna-Prasad-P-V** [comment]
> Hi @Krishna-Prasad-P-V, is this ready for another review?

Yes, it is ready now for another review.

**Comment by yzhang1991** [comment]
@Krishna-Prasad-P-V In the future, it will be great if you can just ping the corresponding reviewers to have another pass. @aditi-pandit

**Comment by aditi-pandit** [comment]
@Krishna-Prasad-P-V : We try to have minimal commits for each PR by separating the PR branch from main and rebasing from it as needed. Please can you try to collapse your changes to a single commit. The changes are looking okay otherwise.

**Comment by Krishna-Prasad-P-V** [comment]
> @Krishna-Prasad-P-V : We try to have minimal commits for each PR by separating the PR branch from main and rebasing from it as needed. Please can you try to collapse your changes to a single commit. The changes are looking okay otherwise.

I have opened another PR (https://github.com/facebookincubator/velox/pull/5759) and closing this one. It is difficult squash so many commits into one and with rebasing being done regularly as the PR has been opened a month before. Hence, closing this PR as another new one with a single commit is available.

**Pull request closed without merging**